Add dir-diff support and various diff options

jhenkens opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->
Looking for feedback on the below changes - I don't think they are ready to merge in, but looking for early advice (or taking over ownership, if there is an eager maintainer).

The dir-diff support is in relation to https://github.com/gitextensions/gitextensions/issues/1647
The diff modes are improvements I wanted while working on the dir-diff support.

Fixes #

Changes proposed in this pull request:
- Add support for diff-tool, in the settings page
- Add two new compare modes 
  - Compare to Working Directory
  - Compare Selected Commits (use CTRL + click to select two commits, compare them in the order selected)

 
Screenshots before and after (if PR changes UI):
![image](https://user-images.githubusercontent.com/242763/40944525-fcd1d19a-6809-11e8-8b88-44be204e544e.png)
![image](https://user-images.githubusercontent.com/242763/40944547-0fdb5392-680a-11e8-8903-e14548eb637f.png)

What did I do to test the code and ensure quality:
- Local testing with my setup


Has been tested on (remove any that don't apply):
- GIT 2.17
- Windows 10

vbjay commented:
Fyi. To signify a pull request as not ready just add WIP to tye beginning
of the title. It stands for work in progress.

On Mon, Jun 4, 2018, 6:38 PM codecov[bot] <notifications@github.com> wrote:

> Codecov
> <https://codecov.io/gh/gitextensions/gitextensions/pull/5025?src=pr&el=h1>
> Report
>
> Merging #5025
> <https://codecov.io/gh/gitextensions/gitextensions/pull/5025?src=pr&el=desc>
> into master
> <https://codecov.io/gh/gitextensions/gitextensions/commit/e532aad6c992ca4c42bb528a9482ae54e1e97bb5?src=pr&el=desc>
> will *decrease* coverage by 0.1%.
> The diff coverage is 0%.
>
> @@            Coverage Diff             @@##           master    #5025      +/-   ##
> ==========================================- Coverage   34.77%   34.67%   -0.11%
> ==========================================
>   Files         563      563
>   Lines       44434    44462      +28
>   Branches     6123     6125       +2
> ==========================================- Hits        15450    15415      -35     - Misses      28214    28291      +77     + Partials      770      756      -14
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5025#issuecomment-394522892>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsdqGoSBkLbrbarClHW3275R_siJTks5t5bb9gaJpZM4UZ4bo>
> .
>

drewnoakes commented:
Looks good at a glance, though I don't know much about diff tools so won't offer to be a primary reviewer.

Can you elaborate on why you feel this isn't ready for merging?

jhenkens commented:
No unit testing, very limited testing of clients, known issue https://github.com/gitextensions/gitextensions/pull/5025/files#diff-05956b8e9b35344894a0ffe609cf32b9R3447 that I'm not certain is a local problem, or a tooling problem.

That issue manifests when clicking Working Directory, then compare to current branch.

gerhardol commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -3644,7 +3648,7 @@ private void CompareToBranchToolStripMenuItem_Click(object sender, EventArgs e)
                 {
                     var baseCommit = Module.RevParse(form.BranchName);
-                    UICommands.ShowFormDiff(IsFirstParentValid(), baseCommit, headCommit.ObjectId,
+                    UICommands.PerformDiff(IsFirstParentValid(), baseCommit, headCommit.ObjectId,
```
Should this really be changed?

jhenkens replied:
I'm not sure the issue here. I presume we would want to be able to perform a directory diff from "Compare to branch" popup dialog, would we not?

Or is the comment in regards to the general renaming from ShowFormDiff to PerformDiff. That was a decision I made to clearly distiguish that this may not show a GitExtensions form. PerformDiff encompases the logic that either launches the dir-diff tool directly, or shows a GitExtensions form.

gerhardol commented on GitCommands/Git/GitModule.cs:
```diff
@@ -3444,6 +3444,15 @@ public IEnumerable<string> GetPreviousCommitMessages(int count, string revision
         }
 
+        // Known bug on git-for-windows: Comparing working directory to any branch, fails, due to -R
+        // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
```
If so, information about the diff need to be transferred to _revisionDiffProvider.Get()
The -R option is used to display diffs correctly when comparing staged/unstaged artificial commits to other commits.

jhenkens replied:
First - I am but one person using this application. I did not know if this "known bug" is actually a bug for all, or a bug for me. I'd appreciate help testing this.

Second, I understand the need for the -R, but I was unsure as to what to do in this case, if the git-for-windows bug is truly broken for everyone. I could add logic to _revisionDiffProvider to know whether or not it is a dir-diff, and open a backwards diff. Or fall back to the GitExtensions display, file by file.

RussKie commented:
Does this bug affect all versions of gfw or only specific ones? Has it been
fixed?
E.g. if it was fixed in the min required version (2.9.0?) then we don't
need to worry about it...


On Tue, Jun 12, 2018, 3:03 AM Johan Henkens <notifications@github.com>
wrote:

> *@jhenkens* commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/GitModule.cs
> <https://github.com/gitextensions/gitextensions/pull/5025#discussion_r194582500>
> :
>
> > @@ -3444,6 +3444,15 @@ public IEnumerable<string> GetPreviousCommitMessages(int count, string revision
>                  });
>          }
>
> +        // Known bug on git-for-windows: Comparing working directory to any branch, fails, due to -R
> +        // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
>
> First - I am but one person using this application. I did not know if this
> "known bug" is actually a bug for all, or a bug for me. I'd appreciate help
> testing this.
>
> Second, I understand the need for the -R, but I was unsure as to what to
> do in this case, if the git-for-windows bug is truly broken for everyone. I
> could add logic to _revisionDiffProvider to know whether or not it is a
> dir-diff, and open a backwards diff. Or fall back to the GitExtensions
> display, file by file.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5025#discussion_r194582500>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXl7MWW9yJSGsfdHXG1aiskefg_efks5t7wVbgaJpZM4UZ4bo>
> .
>

jhenkens commented:
I guess I'll clarify. It's a bug that I have found, but not found any other mentions of. I don't think its known to gfw developers as of yet. I've reproduced it on two systems, one with GFW 2.17.0, another with 2.17.1.windows.2

jhenkens commented:
Unless I hear otherwise, I think the best course of action is #2 - "disable comparing "From" the working directory when --dir-diff is enabled. One could still compare "to" the working directory, via the new "Compare to Working Directory" button, or via "Select as base" - "Compare to base"".

I will tackle this later this week and clean everything else up.

jhenkens commented:
Any further thoughts or blockers?

gerhardol commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
+
+            compareToBranchToolStripMenuItem.Enabled = enabledDiffOnWorkingDirectory;
+            compareWithCurrentBranchToolStripMenuItem.Enabled = enabledDiffOnWorkingDirectory;
+            compareSelectedCommitsMenuItem.Enabled = enabledDiffOnWorkingDirectory;
```
only use check above. Exception with builtin diff if only one is selected

jhenkens replied:
Bug. Fixed.

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/DiffViewerSettingsPage.Designer.cs:
```diff
@@ -51,6 +52,7 @@ private void InitializeComponent()
             this.tableLayoutPanelForDiffViewer.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
             this.tableLayoutPanelForDiffViewer.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
             this.tableLayoutPanelForDiffViewer.Controls.Add(this.chkShowDiffForAllParents, 0, 6);
+            this.tableLayoutPanelForDiffViewer.Controls.Add(this.useDifftoolDirDiff, 0, 8);
```
This is not really related to the diff viewer, but there are not really better place for it. Advanced/Detailed need to be rearranged...

jhenkens replied:
Agreed, feels weird here, but didn't change.

gerhardol commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
+            // Thus, we disable comparing "from" working directory.
+            var enabledDiffOnWorkingDirectory = !(AppSettings.UseDifftoolDirDiff && firstSelectedRevision?.Guid == GitRevision.UnstagedGuid);
+
+            compareSelectedCommitsMenuItem.Enabled = selectedRevisions.Count == 2 && firstSelectedRevision != null && secondSelectedRevision != null && enabledDiffOnWorkingDirectory;
```
Most behavior (at least in this menu) is allowing more than on revisions when selecting (the last two clicked are used). This changes this slightly

jhenkens replied:
I believe I've changed this to work as you mention.

jhenkens commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -3766,10 +3781,34 @@ private void compareToBaseToolStripMenuItem_Click(object sender, EventArgs e)
+            var baseCommit = revisions.Skip(1)
+                .First();
+
+            UICommands.PerformDiff(IsFirstParentValid(), baseCommit.ObjectId, headCommit.ObjectId,
```
See above.

jhenkens commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -3766,10 +3781,34 @@ private void compareToBaseToolStripMenuItem_Click(object sender, EventArgs e)
+                return;
+            }
+
+            UICommands.PerformDiff(IsFirstParentValid(), baseCommit.ObjectId, ObjectId.UnstagedId,
```
This was an error about trying to find a merge-base for the compare to merge base checkbox. I made that conditional based on whether or not a merge base could be found, and disabled the checkbox if a merge base could not.

jhenkens commented on GitUI/UserControls/RevisionGrid.cs:
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
             if (selectedRevisions.Count == 1 && firstSelectedRevision != null)
             {
                 _navigationHistory.Push(firstSelectedRevision.Guid);
+                compareToWorkingDirectoryMenuItem.Enabled = firstSelectedRevision.ObjectId != ObjectId.UnstagedId;
```
Fixed.

jhenkens commented:
Missed the general feedback.

>Some comments about details, but I have a more general comment: I would expect dir-diff to be available together with FormDiff, not either or. Maybe the new menu options should be dedicated to dir-diff. (With the updates to artificial revisions, I hardly use FormDiff though.)

I think that they both provide essentially the same functionality in two very separate user interfaces. As someone who is quite used to a "dir-diff" workflow, I'd say having them comingled, or a second button click away (from within the "FormDiff") window adds a good bit of extra clicks needed to get what is desired, done.

jhenkens commented:
@gerhardol I think I have addressed your concerns and responded to your overall discussion. I have just merged up with the latest refactors that have happened on default, and reapplied some changes that were initially in conflict. Please let me know what you think

jhenkens commented:
Ah - I do now see a flaw with the "getting rid of the form-diff" window. It also eliminates the "compare to merge base" possibility, which is actually quite critical a lot of the time when diffing.

I'd love some suggestions on how to do this elegantly, from anyone involved. I feel the combination of:
`Right Click -> Dropdown -> Click` is already somewhat heavy (but I'll accept it as its what is already here).

But `Right` Click -> Dropdown -> Click -> Click Checkbox -> Click Open in directory diff tool` just seems horrendous.

jhenkens commented:
Added customizable shortcuts to the dropdowns for compare, and also added mnemonics to the form diff buttons, so that one can easily navigate the compare workflows I've describe entirely via keyboard. 

I think the changeset is now in a good spot.

gerhardol commented on GitUI/CommandsDialogs/FormDiff.cs:
```diff
@@ -82,6 +86,13 @@ private void PopulateDiffFiles()
+            // Bug in git-for-windows: Comparing working directory to any branch, fails, due to -R
+            // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
+            // git difftool --gui --no-prompt --dir-diff HEAD succeeds
+            // Thus, we disable comparing "from" working directory.
```
Open with difftool button is still enabled

jhenkens replied:
Yes - the issue was `            openWithDifftoolToolStripMenuItem.Enabled = enableDifftoolDirDiff;` rather than `            btnCompareDirectoriesWithDiffTool.Enabled = enableDifftoolDirDiff;`. A bug.

gerhardol commented on GitUI/UserControls/RevisionGrid/RevisionGridControl.Designer.cs:
```diff
@@ -232,7 +235,6 @@ private void InitializeComponent()
             // 
             this.createNewBranchToolStripMenuItem.Image = global::GitUI.Properties.Images.BranchCreate;
             this.createNewBranchToolStripMenuItem.Name = "createNewBranchToolStripMenuItem";
-            this.createNewBranchToolStripMenuItem.ShortcutKeys = ((System.Windows.Forms.Keys)((System.Windows.Forms.Keys.Control | System.Windows.Forms.Keys.B)));
```
Note: Not in this PR, the shortcut keys removed in this file are not in Settings

jhenkens replied:
This shortcut still works without issue from the menu-bar hotkey, which is reassignable via the settings page.

jhenkens commented on GitUI/CommandsDialogs/FormDiff.cs:
```diff
@@ -219,6 +230,16 @@ private void ckCompareToMergeBase_CheckedChanged(object sender, EventArgs e)
 
+        private void setCkCompareToMergeBaseText(string commit)
+        {
+            ckCompareToMergeBase.Text = $"Compare to merge &base ({commit})";
```
I made the change, but it eliminates the mnemonic, sadly. Not sure why, but don't think it was worth my time.

jhenkens commented:
I don't see a good pattern for ways to test this based on existing tests.

The closest existing tests I could find were for the GitCommands. Testing the git difftool --dir-diff command is difficult - it opens an external tool, only. Not sure its a great candidate. The rest are function UI tests, the closest equivalents I could find were model tests on a few helper models.

gerhardol commented:
> I don't see a good pattern for ways to test this based on existing tests.

@RussKie 
Comments on tests here?
I do not believe that tests add value for these changes (other than increasing the reported test coverage).

RussKie commented:
Please rebase on top of the current master and be ready to squash.

RussKie commented:
![image](https://user-images.githubusercontent.com/4403806/43700654-c370eba2-9996-11e8-9867-65656e6580c2.png)

I guess P4merge isn't capable of comparing directories... 😯

RussKie commented on GitUI/CommandsDialogs/FormDiff.Designer.cs:
```diff
@@ -166,15 +170,39 @@ private void InitializeComponent()
+            // btnCompareDirectoriesWithDiffTool
+            // 
+            this.btnCompareDirectoriesWithDiffTool.Anchor = AnchorStyles.Left;
+            this.btnCompareDirectoriesWithDiffTool.AutoSize = true;
```
Need to add `AutoSizeMode = GrowAndShrink`

jhenkens replied:
✔️

gerhardol commented:
> I guess P4merge isn't capable of comparing directories...

Beyond Compare is good at all compares.
For free applications, at least WinMerge (at least betas) support this.

drewnoakes commented:
Meld is a decent FOSS option.

jhenkens commented on GitUI/CommandsDialogs/FormDiff.cs:
```diff
@@ -219,6 +231,16 @@ private void ckCompareToMergeBase_CheckedChanged(object sender, EventArgs e)
             PopulateDiffFiles();
         }
 
+        private void setCkCompareToMergeBaseText(string commit)
```
✔️

jhenkens commented:
@RussKie Perhaps I should add back in the options page, to allow a user to checkbox whether or not their diff tool supports directory diffing?

jhenkens commented on GitUI/CommandsDialogs/FormDiff.Designer.cs:
```diff
@@ -166,15 +170,39 @@ private void InitializeComponent()
+            this.btnCompareDirectoriesWithDiffTool.Name = "btnCompareDirectoriesWithDiffTool";
+            this.btnCompareDirectoriesWithDiffTool.Margin = new System.Windows.Forms.Padding(0);
+            this.btnCompareDirectoriesWithDiffTool.Size = new System.Drawing.Size(141, 17);
+            this.btnCompareDirectoriesWithDiffTool.TabIndex = 20;
```
New to winforms - is there anything TabIndex does besides set the order over movement via cursor for tabs? I notice TabIndex set on a lot of components that are not selectable (panels, groups, etc), as well as the main diff panel, which once entered you cannot leave via tab. To me, the tab index on this entire page feels weird, but maybe its a PEBKAC.

vbjay commented:
Click view tab order when in the form designer. You'll get the idea pretty
quickly.

On Wed, Aug 8, 2018, 6:49 PM Johan Henkens <notifications@github.com> wrote:

> *@jhenkens* commented on this pull request.
> ------------------------------
>
> In GitUI/CommandsDialogs/FormDiff.Designer.cs
> <https://github.com/gitextensions/gitextensions/pull/5025#discussion_r208760096>
> :
>
> >              this.ckCompareToMergeBase.UseVisualStyleBackColor = true;
>              this.ckCompareToMergeBase.CheckedChanged += new System.EventHandler(this.ckCompareToMergeBase_CheckedChanged);
>              //
> +            // btnCompareDirectoriesWithDiffTool
> +            //
> +            this.btnCompareDirectoriesWithDiffTool.Anchor = AnchorStyles.Left;
> +            this.btnCompareDirectoriesWithDiffTool.AutoSize = true;
> +            this.btnCompareDirectoriesWithDiffTool.Name = "btnCompareDirectoriesWithDiffTool";
> +            this.btnCompareDirectoriesWithDiffTool.Margin = new System.Windows.Forms.Padding(0);
> +            this.btnCompareDirectoriesWithDiffTool.Size = new System.Drawing.Size(141, 17);
> +            this.btnCompareDirectoriesWithDiffTool.TabIndex = 20;
>
> New to winforms - is there anything TabIndex does besides set the order
> over movement via cursor for tabs? I notice TabIndex set on a lot of
> components that are not selectable (panels, groups, etc), as well as the
> main diff panel, which once entered you cannot leave via tab. To me, the
> tab index on this entire page feels weird, but maybe its a PEBKAC.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5025#discussion_r208760096>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsciHd2REAmGTl3Z6HErvMbRSk2zwks5uO2r0gaJpZM4UZ4bo>
> .
>

jhenkens commented:
Ah - the things I didn't know, as a Rider based developer. I'll give it a shot.

gerhardol commented:
> Last question is posted a few comments - re the options page.

The diff-tool popup is good enough for me - the user should know that they have tried a directory diff and decipher the error message. There are enough options in GE already (even if I personally like to be able to change details). The diff/merge tool settings should be rewritten anyway. I vote for no option.

That said, I fully understand if someone else request the option.

jhenkens commented:
Hate to be that-guy, but ping?

Any other comments?

vbjay commented:
![image](https://user-images.githubusercontent.com/3629489/44118905-8f70d740-9fe5-11e8-8046-2893d5d80357.png)

jhenkens commented:
![image](https://user-images.githubusercontent.com/242763/44119935-92e6489e-9fcf-11e8-891f-d1d2fbc04b6c.png)

The only commoent "see review" sends me to _was addressed by me_. Everything else from that review was also addressed.

![image](https://user-images.githubusercontent.com/242763/44119983-b7caaf1a-9fcf-11e8-9d22-48a7a988ae5e.png)

As well as updating the tab indexes via Visual Studio to make sense.

RussKie commented:
Apologies, I haven't forgotten about it. I'm absolutely flat out busy with
family and work commitments at present.
I'll review the changes as soon as I can.

On Wed, Aug 15, 2018, 7:39 AM Johan Henkens <notifications@github.com>
wrote:

> [image: image]
> <https://user-images.githubusercontent.com/242763/44119935-92e6489e-9fcf-11e8-891f-d1d2fbc04b6c.png>
>
> The only commoent "see review" sends me to *was addressed by me*.
> Everything else from that review was also addressed.
>
> [image: image]
> <https://user-images.githubusercontent.com/242763/44119983-b7caaf1a-9fcf-11e8-9d22-48a7a988ae5e.png>
>
> As well as updating the tab indexes via Visual Studio to make sense.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/5025#issuecomment-413025203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXpP3tVDxoa68E9yH1iyLRv_rMJUTks5uQ0OhgaJpZM4UZ4bo>
> .
>

RussKie commented:
Thank you
